site/content/languages/zig.md

```diff
@@ -0,0 +1,213 @@
+the heap base. All memory beyond that is used for the Zig heap.
+
+The Zig language performs no memory management on behalf of the programmer. 
+However, Zig has no default allocator. Instead, functions which need to allocate 
+accept an `Allocator` parameter.
+
+### Host Allocations
+
+Sometimes a host function needs to allocate memory directly. For example, to write JSON
+of a given length before invoking an exported function to parse it.
+
+```zig
+pub export fn configure(ptr: [*]const u8, size: u32) void {
+    _configure(message[0..size]) catch |err| @panic(switch (err) {
+        error.OutOfMemory => "out of memory",
+    });
+}
+```
+
+The general flow is that the host allocates memory by calling an allocation
+function with the size needed. Then, it writes data, in this case JSON, to the
+memory offset (`ptr`). At that point, it can call a host function, ex
+`configure`, passing the `ptr` and `size` allocated. The guest Wasm (compiled
+from Zig) will be able to read the data. To ensure no memory leaks, the host
+calls a free function, with the same `ptr`, afterwards and unconditionally.
+
+Note: wazero includes an [example project][9] that shows this.
+
+The [zig example][9] does a few things of interest:
+* Uses `@ptrToInt` to change a Zig pointer to a numeric type
+* Uses `[*]u8` as an argument to take a pointer and slices it to build back a
+string
+* It also shows how to import a host function using the `extern` directive
+
+To allow the host to allocate memory, you need to define your own `malloc` and
+`free` functions:
+```webassembly
+(func (export "malloc") (param $size i32) (result (;$ptr;) i32))
+(func (export "free") (param $ptr i32) (param $size i32))
+```
+
+Because Zig easily allows end-users to [plug their own allocators][12], it relatively easy to 
+export custom `malloc`/`free` pairs to the host.
+
+For instance, the following code exports `malloc`, `free` from Zig's `page_allocator`:
+
+```zig
+const allocator = std.heap.page_allocator;
+
+pub export fn malloc(length: usize) ?[*]u8 {
+    const buff = allocator.alloc(u8, length) catch return null;
+    return buff.ptr;
+}
+
+pub export fn free(buf: [*]u8, length: usize) void {
+    allocator.free(buf[0..length]);
+}
+```
+
+## System Calls
+
+Please read our overview of WebAssembly and
+[System Calls]({{< ref "_index.md#system-calls" >}}). In short, WebAssembly is
+a stack-based virtual machine specification, so operates at a lower level than
+an operating system.
+
+For functionality the operating system would otherwise provide, you must use
+the `wasm32-wasi` target. This imports host functions in
+[WASI]({{< ref "/specs#wasi" >}}).
+
+For example, `zig build-exe -target wasm32-wasi hello.zig` compiles the
+below `main` function into a WASI function exported as `_start`.
+
+```zig
+const std = @import("std");
+const io = std.io;
+const os = std.os;
+const allocator = std.heap.page_allocator;
```

Comment:
I think allocator is nothing to do with system calls (they do not make any system calls).